pandas/core/internals.py

```diff
@@ -492,6 +492,14 @@ def _astype(self, dtype, copy=False, errors='raise', values=None,
                 return self.copy()
             return self
 
+        # numpy arrays don't handle generic timestamp dtypes well. Since
+        # we only use the nanosecond frequency, interpret generic timestamp
+        # dtypes as nanosecond frequency.
+        if dtype.name == "datetime64":
```

Comment:
I would agree, except that when we call `.astype`, `maybe_cast_to_datetime` does not get called.  But certainly, there are other places that casting can go.